### PR TITLE
fix: Timeout crash because of chunked_output len

### LINKED ISSUE #2554: Chunked output causes timeout crash on certain seq len
# TL;DR

If one of dataloader batches is 49 tokens long, torchtune crashes on timeout

# Longer explanation

`chunked_output` in transformer.py splits output into a list of 8 tensors.
If output of length 49 is expected, transformer outputs a list where first 7 tensors have length 7 but the last one has length 0 - that's the main issue. FSDP all_gather func waits for all 8 tensors to sync gradients but never gets the 8th one because it's empty.
Same works with lengths 7, 14, 21, 28, 35, 42.

- The issue is not reproducible if output layer is not sharded
- The issue is not reproducible if all microbatches have one of broken size - because noone waits for the 8th tensor

# How to reproduce it

In one of recipes, i.e. `full_finetune_distributed` at the main dataloader cycle manually truncate batch for some ranks like that:

```
for idx, batch in enumerate(self._dataloader):
  if torch.distributed.get_rank() % 2 == 0:
    batch["tokens"] = batch["tokens"][:, :49]
    batch["labels"] = batch["tokens"][:, :49]
```

Run it on one of llama3_1 recipes (8B / 70B). You'll get 0th iteration training stuck and a timeout in 30 mins

### PR DESCRIPTION
**Pull request by bogdansalyp**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses - closes #2554 

#### Changelog
What are the changes made in this PR?
* replaced `torch.chunk` with `torch.tensor_split`
#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Review by pbontrager — COMMENTED** [review]
Why not just use torch.tensor_split? Also, we should test what happens in the case of 7 tokens, tensor split will return an 8th empty tensor.

**Comment by bogdansalyp** [comment]
> Why not just use torch.tensor_split? Also, we should test what happens in the case of 7 tokens, tensor split will return an 8th empty tensor.

image

Wow you're right, I didn't know this func differs. Will remove my module from the PR and replace with this func
Regarding splitting 7 into 8 - it gives an empty tensor as the 8th element. Might work out of box, will check tomorrow

**Comment by bogdansalyp** [comment]
@pbontrager @felipemello1 review pls

Added `tensor_split`, added tests for transformer `chunked_output` func (covered simple case, 49 len and 7 len into 8 chunks)

**Review by felipemello1 — COMMENTED** [review]
LGTM, thank you! Just left one comment.

Is there any chance that you can run it with compile=True for just a few steps (~10) and share the run stats from wandb (or some other logger), to make sure things look fine?

File: torchtune/modules/loss/ce_chunked_output_loss.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def forward(self, logits: List[torch.Tensor], labels: torch.Tensor) -> torch.Ten
             >>> loss_fn = ChunkedCrossEntropyLoss()
             >>>
             >>> h = torch.tensor([bsz, num_tokens, dim])
-            >>> output_chunks = [model.output(chunk) for chunk in h.chunk(num_chunks, dim=1)]
+            >>> output_chunks = [model.output(chunk) for chunk in h.tensor_split(hum_chunks, dim=1)]
```
[the remark this answers is not in the document]
Comment (bogdansalyp) [reply torchtune/modules/loss/ce_chunked_output_loss.py]: fixed

File: tests/torchtune/modules/test_transformer_decoder.py [hunk 2]
```diff
@@ -369,6 +378,45 @@ def test_forward(
...
+            )
+
+    @mps_ignored_test()
+    def test_forward_output_chunks_corner_case(
+        self,
+        input_chunked_corner_case: torch.Tensor,
+        input_params: Tuple[int, int, int],
+        decoder: TransformerDecoder,
+    ) -> None:
+        """Checks chunked output corner case (torch.chunk would've split in 7)."""
```
[the remark this answers is not in the document]
Comment (bogdansalyp) [reply tests/torchtune/modules/test_transformer_decoder.py]: renamed funcs and expanded docstring

**Comment by bogdansalyp** [comment]
image
image


launched with `compile=True`

**Comment by bogdansalyp** [comment]
@felipemello1 addressed your change requests & attached wandb

**Review by felipemello1 — APPROVED** [review]
